### ISSUE TITLE
Reorder Feature Matrix by increasing number of channels

### ISSUE THREAD

**Issue by prakharg24** [PaddlePaddle/Paddle#19550]
Hi,

I want to do reordering of a feature matrix in my CNN model, such that I can decrease the size of the feature matrix by increasing its number of channels, without losing any information.

For example,
300 x 300 x 3 --> 150 x 150 x 12

The reason that I cannot simply use resize is because it distributes the data in a different way that I want it to.

Let's try to convert this ->
Channel 1->
1 2 3 4
5 6 7 8
9 0 a b
c d e f

What resize will do ->
Channel 1->
1 2
5 6
Channel 2->
3 4
7 8
and so on...

What I want
Channel 1 ->
1 3
9 a
Channel 2->
2 4
0 b
and so on...

1. Am I understanding correctly what resize will do?
2. Is there some other function that can do what I want?
3. If there is no other direct function for it, how do I implement this layer for my CNN?

Thank You

**Comment by qingqing01** [comment]
@prakharg24 You can try to use `reshape`  and `tranpose` to comb

**Comment by prakharg24** [comment]
@qingqing01 I haven't been able to figure out exactly how...

Lets say you have input=x with shape -1 x 3 x 300 x 300 (-1 is for batch size)
you want y with shape -1 x 12 x 150 x 150 with the reordering as mentioned above..

How exactly will you do it using reshape and transpose??

**Comment by prakharg24** [comment]
I need to convert -1 x c x 150 x 150 to -1 x c*16 x 38 x 38

So, I am going with this ->
```
        module10_pad = fluid.layers.pad2d(module10, [0, 2, 0, 2])
        module10_reo = fluid.layers.reshape(module10_pad, [-1, 1, 152, 152])
        module10_reo = fluid.layers.reshape(module10_reo, [-1, 38, 4, 152])
        module10_reo = fluid.layers.transpose(module10_reo, perm=[0, 1, 3, 2])
        module10_reo = fluid.layers.reshape(module10_reo, [-1, 38*38, 4, 4])
        module10_reo = fluid.layers.transpose(module10_reo, perm=[0, 1, 3, 2])
        module10_reo = fluid.layers.reshape(module10_reo, [-1, 38*38, 4*4])
        module10_reo = fluid.layers.transpose(module10_reo, perm=[0, 2, 1])
        module10_reo = fluid.layers.reshape(module10_reo, [-1, 38*38])
        module10_reo = fluid.layers.reshape(module10_reo, [-1, 38, 38])
```

I experimented with the same pattern of reshape and transpose in numpy and it worked the way I wanted it to, so I am assuming similar behaviour will be seen for fluid.layers.reshape and transpose too.

I am confused as to whether I should include the batch axis in reshape and transpose or just ignore it completely (like in tensorflow). I ignored it in the code above. Is that the right way to go?

**Comment by luotao1** [comment]
You can refer to the [reshape API](https://www.paddlepaddle.org.cn/documentation/docs/en/api/layers/nn.html#reshape) at first.

**Comment by phlrain** [comment]
@prakharg24 sorry about my poor english， If you can read this,  I will translate it for you..
我觉得你的困惑可能是在 layers.data定义的tensor ，在什么时候和我指定的shape不一致，
比如你定义了 layers.data( "a", shape=[10, 10])，什么时候就变成了[-1, 10, 10]
现在的逻辑是这样的，如果你的shape里面没有-1，并且是用来 默认值 append_batch_size=true, 这个shape会变成 [-1, 10, 10], rank值会加 1

如果你的shape里面包含了-1, 无论是否设置append_batch_size 的值，shape都不会变

如果你想表现的和tf一样，每次定义的时候 ，直接设置append_batch_size=False

目前这种设计却是给用户带来了困扰，们已经在实现一种新的接口方式，不会做任何隐式的变换和操作，就跟tf 和torch一样

**Comment by prakharg24** [comment]
@phlrain okay so to put it in simpler terms

My input tensor ->
```
shape(X) = -1, 64, 152, 152
```
if I do 
```
Y = fluid.layers.reshape(X, [-1, 1, 152, 152])
```

then what will be shape(Y)?
If I was using tensorflow, then I would have gotten
```
shape(Y) = -1, 64, 1, 152, 152
```

Is it the same for paddlepaddle? Or something else will happen?

**Comment by phlrain** [comment]
![image](https://user-images.githubusercontent.com/43953930/64016925-ddc9dd00-cb5a-11e9-9cde-e41e719a61ff.png)
I test some code in tf. 
shape(x) = -1, 64, 152, 152

**Comment by phlrain** [comment]
![image](https://user-images.githubusercontent.com/43953930/64016973-03ef7d00-cb5b-11e9-893d-580411ee8280.png)

I test some code in TF

shape(x) = -1, 64, 152, 152
y = tf.reshape( x, [-1, 1, 152, 152])
shape(y) =  -1, 1, 152, 152

paddle is same with tf

@prakharg24

**Comment by prakharg24** [comment]
@phlrain I understood what I was doing wrong.

To convert a tensor of shape -1, 64, 152, 152 to -1, 64, 1, 152, 152 I have to do the following,
```
y = fluid.layers.reshape(x, [0, 64, 1, 152, 152]
```

Basically, I need to add a '0' for the batch axis in reshape and also consider it during transpose.

Thank You

**Comment by phlrain** [comment]
@prakharg24

I think you want to something like this.
y= fluid.layers.reshape( x, [ layers.shape(x)[0], 64, 1, 152, 152] )

paddle can use like this. 

0 in shape will Deprecated in the future

**Comment by prakharg24** [comment]
@phlrain  oh okay. Thanks. Will keep this in mind
